include/eve/arch/riscv/as_register.hpp

```diff
@@ -0,0 +1,187 @@
+      if constexpr( lmul == 2 ) return find_##name_part##m2();                                     \
+      if constexpr( lmul == 4 ) return find_##name_part##m4();                                     \
+      if constexpr( lmul == 8 ) return find_##name_part##m8();                                     \
+      if constexpr( lmul == -2 ) return find_##name_part##mf2();                                   \
+      if constexpr( lmul == -4 ) return find_##name_part##mf4();                                   \
+      FRAC_LMUL_8(if constexpr( lmul == -8 ) return find_##name_part##mf8();)                      \
+    }
+
+#  define FRAC_LMUL_8(X) X
+  EVE_RISCV_LMUL_VARIANT(vint)
+  EVE_RISCV_LMUL_VARIANT(vuint)
+#  undef FRAC_LMUL_8
+#  define FRAC_LMUL_8(X)
+  EVE_RISCV_LMUL_VARIANT(vfloat)
+#  undef FRAC_LMUL_8
+#  undef EVE_RISCV_LMUL_VARIANT
+
+  public:
+  static consteval auto find()
+  {
+    constexpr auto width = sizeof(Type) * Size::value * 8;
+
+    static_assert(width <= __riscv_v_fixed_vlen * ABI::MaxLmul,
+                  "[eve riscv] - Type is not usable in SIMD register (too big)");
+
+    if constexpr( is_fp_v ) return find_vfloat();
+    if constexpr( !is_fp_v && is_signed_v ) return find_vint();
+    if constexpr( !is_fp_v && !is_signed_v ) return find_vuint();
+  }
+  using type = decltype(find());
+  static_assert(!std::is_void_v<type>,
+                "[eve riscv] - Type is not usable in a SIMD register (unknown type)");
+};
+
+// ---------------------------------------------------------------------------------------------
+// logical cases
+template<typename Type, typename Size, rvv_abi ABI> struct as_logical_register<Type, Size, ABI>
+{
+  static constexpr auto   lmul     = riscv_rvv_dyn_::getLMUL<Type>(Size::value);
+  static constexpr size_t size     = sizeof(Type) * 8;
+  static constexpr size_t bit_size = lmul > 0 ? size / lmul : size * (-lmul);
+
+#  ifdef EVE_RVV_RET_MASK_TYPE
+#    error EVE_RVV_RET_MASK_TYPE defined
+#  endif
+#  define EVE_RVV_RET_MASK_TYPE(bit_size)                                                          \
+    using type = vbool##bit_size##_t                                                               \
+        __attribute__((riscv_rvv_vector_bits(__riscv_v_fixed_vlen / bit_size)));                   \
+    return type                                                                                    \
+    {}
+
+  static constexpr auto find()
+  {
+    if constexpr( bit_size == 1 ) { EVE_RVV_RET_MASK_TYPE(1); }
+    if constexpr( bit_size == 2 ) { EVE_RVV_RET_MASK_TYPE(2); }
+    if constexpr( bit_size == 4 ) { EVE_RVV_RET_MASK_TYPE(4); }
+    if constexpr( bit_size == 8 ) { EVE_RVV_RET_MASK_TYPE(8); }
+    if constexpr( bit_size == 16 ) { EVE_RVV_RET_MASK_TYPE(16); }
+    if constexpr( bit_size == 32 ) { EVE_RVV_RET_MASK_TYPE(32); }
+    if constexpr( bit_size == 64 ) { EVE_RVV_RET_MASK_TYPE(64); }
+  }
```

Comment:
heavy macro usage makes this very difficult to review.

How close can you get to?

```
template <typename rvv_type, std::size_t num, std::size_t denum>
using rvv_with_attr = rvv_type __attribute__((riscv_rvv_vector_bits(__riscv_v_fixed_vlen * num / denum)));

auto find() {
   if constexpr ( std::same_as<T, std::int8_t> ) {
      if constexpr ( N::value == 1 ) return rvv_with_attr<...>{}; 
   }
}
```

Reply:
Removed defines from `as_register.hpp`.